cpp/include/cudf/io/orc.hpp

```diff
@@ -0,0 +1,832 @@
+   */
+  orc_reader_options() = default;
+
+  /**
+   * @brief Creates `orc_reader_options_builder` which will build `orc_reader_options`.
+   *
+   * @param src Source information to read orc file.
+   * @return Builder to build reader options.
+   */
+  static orc_reader_options_builder builder(source_info const& src);
+
+  /**
+   * @brief Returns source info.
+   */
+  source_info const& get_source() const { return _source; }
+
+  /**
+   * @brief Returns names of the columns to read.
+   */
+  std::vector<std::string> const& get_columns() const { return _columns; }
+
+  /**
+   * @brief Returns list of individual stripes to read.
+   */
+  std::vector<size_type> const& get_stripes() const { return _stripes; }
+
+  /**
+   * @brief Returns number of rows to skip from the start.
+   */
+  size_type get_skip_rows() const { return _skip_rows; }
+
+  /**
+   * @brief Returns number of row to read.
+   */
+  size_type get_num_rows() const { return _num_rows; }
+
+  /**
+   * @brief Whether to use row index to speed-up reading.
+   */
+  bool is_enabled_use_index() const { return _use_index; }
+
+  /**
+   * @brief Whether to use numpy-compatible dtypes.
+   */
+  bool is_enabled_use_np_dtypes() const { return _use_np_dtypes; }
+
+  /**
+   * @brief Returns timestamp type to which timestamp column will be cast.
+   */
+  data_type get_timestamp_type() const { return _timestamp_type; }
+
+  /**
+   * @brief Whether to convert decimals to float64.
+   */
+  bool is_enabled_decimals_as_float64() const { return _decimals_as_float64; }
+
+  /**
+   * @brief Returns whether decimal scale is inferred or forced to have limited fractional digits.
+   */
+  size_type get_forced_decimals_scale() const { return _forced_decimals_scale; }
+
+  // Setters
+
+  /**
+   * @brief Sets names of the column to read.
+   *
+   * @param col_names Vector of column names.
+   */
+  void set_columns(std::vector<std::string> col_names) { _columns = std::move(col_names); }
+
+  /**
+   * @brief Sets list of individual stripes to read.
+   *
+   * @param stripes Vector of individual stripes.
+   */
+  void set_stripes(std::vector<size_type> stripes)
+  {
+    CUDF_EXPECTS(stripes.empty() or (_skip_rows == 0), "Can't set stripes along with skip_rows");
+    CUDF_EXPECTS(stripes.empty() or (_num_rows == -1),  "Can't set stripes along with num_rows");
+    _stripes = std::move(stripes);
+  }
+
+  /**
+   * @brief Sets number of rows to skip from the start.
+   *
+   * @param rows Number of rows.
+   */
+  void set_skip_rows(size_type rows)
+  {
+    CUDF_EXPECTS(rows == 0 or _stripes.empty(), "Can't set both skip_rows along with stripes");
+    _skip_rows = rows;
+  }
+
+  /**
+   * @brief Sets number of row to read.
+   *
+   * @param nrows Number of rows.
+   */
+  void set_num_rows(size_type nrows)
+  {
+    CUDF_EXPECTS(nrows == -1 or _stripes.empty(), "Can't set both num_rows along with stripes");
+    _num_rows = (nrows != 0) ? nrows : -1;
```

Comment:
Why is this line required? Parquet reader options don't have this logic.

Reply:
To take care of this condition https://github.com/rapidsai/cudf/blob/c7e8a5b431ff03bb456c8aa54456e40c19c73799/cpp/src/io/orc/reader_impl.cu#L852

Reply:
Looks like the behavior is to read all rows when num_rows is set to zero. I don't think we want this. 
AFAIK this is also true for the Parquet reader, so we probably want to open a separate issue for this. Thoughts?

Reply:
Agree, will remove this and see if any thing breaks.

Reply:
Just checked, and Parquet reads zero rows with zero num_rows (which is correct). So it would be great to align ORC reader in this PR.

In nothing breaks, add a test that would :)